casper/src/main/scala/coop/rchain/casper/blocks/merger/CasperMergingDagReader.scala

```diff
@@ -0,0 +1,54 @@
+package coop.rchain.casper.blocks.merger
+
+import cats.effect.Concurrent
+import cats.syntax.all.{none, _}
+import coop.rchain.blockstorage.BlockStore
+import coop.rchain.blockstorage.syntax._
+import coop.rchain.dag.DagReader
+import coop.rchain.models.BlockHash.BlockHash
+
+final class CasperMergingDagReader[F[_]: Concurrent: BlockStore](
+    hashDAGReader: DagReader[F, BlockHash]
+) extends DagReader[F, MergingVertex] {
+  override def children(vertex: MergingVertex): F[Option[Set[MergingVertex]]] =
+    hashDAGReader.children(vertex.blockHash).flatMap {
```

Comment:
Both of these member functions only use `vertex` to get the `blockHash`. So this can be simply viewed as function composition.
We are starting from these functions (children, parents) with the type roughly like this `cpHash : Hash -> Opt[Hashes]`. They are functions on DagReader which is omitted as the first parameter.

What we need is `cpVertex : Vertex -> Opt[Vertexes]` which is like lifted function from Hash to Vertex.

Let's see what we already have. `vxHash : Vertex -> Hash` and `cpHV : Hashes -> Opt[Vertexes]` (now inside match).

We can compose it `cpVertex = vxHash andThen cpHash andThen cpHv`. For careful eye, of course there is one problem, `cpHash` returns `Opt[Hashes]` so it must be `flatMap`-ed. :exclamation: **No matching needed. Matching is already done in implementation of Monad.**

In this strange notation the types goes like this.
`cpVertex : Vertex -> Hash -> Opt[Hashes].flatMap( Hashes -> Opt[Vertexes] )`

NOTE: because we are in Scala class may be useful to group implicit dependencies and create extension methods. But this will be much easier in Scala 3. :partying_face: